src/hotspot/share/runtime/objectMonitor.inline.hpp

```diff
@@ -102,6 +102,12 @@ inline void ObjectMonitor::add_to_contentions(int value) {
   Atomic::add(&_contentions, value);
 }
 
+inline void ObjectMonitor::set_recursions(size_t recursions) {
+  assert(_recursions == 0, "must be");
```

Comment:
Why have the `recursions` parameter if the passed value must always be zero?

Update: It looks like you might be trying to detect some out of sync count coming
from the removal of the object from the lock stack. You expect it to always be a
count value of 1 removal and if more than 1 is removed you want to assert().

Reply:
The assert is on the pre-value of `ObjectMonitor::_recursions`. The parameter can have any value. 
Maybe the assert is out of place. 
From the recursive lightweight's perspective `ObjectMonitor::set_recursions` is part of the ObjectMonitors initialisation. Some other thread has created the ObjectMonitor for the locking thread, but filled in the placeholder values anonymous owner and set recursions to 0. The locking thread will at some point notice this and finish initialising the ObjectMonitor by setting the _owner and _recursions fields. There are also assumptions in the C2 code that the _recursions field when inflated by a non owning thread is set to 0.

Reply:
My mistake. I misread that the assertion was against `_recursions` and not `recursions`.